### PR TITLE
Relax `eos_token_id < 0` checks in `generate()` from `ValueError` to warning

### PR DESCRIPTION
**Pull request by lewtun**
# What does this PR do?

This PR relaxes a constraint that was introduced in eec46b4 to ensure `eos_token_id > 0` when `min_new_tokens > 0`. In particular, the current code would throw a `ValueError`:

```python
from transformers import pipeline

pipe = pipeline("text-generation")
pipe("Hello, my dog is cute.", min_new_tokens=4, eos_token_id=-1)
```

This isn't ideal as setting `eos_token_id=-1` is a handy trick to enable generating text past the EOS token. The current PR now returns a warning instead. I've also taken the liberty to relax the error on `min_length < 0` but let me know if that's not a good idea.

Internal Slack thread where this was discussed: https://huggingface.slack.com/archives/C01N44FJDHT/p1680175455762649

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Review by gante — APPROVED** [review]
Thanks for fixing it 🙏

File: src/transformers/generation/logits_process.py [hunk 1]
```diff
@@ -106,12 +107,12 @@ class MinLengthLogitsProcessor(LogitsProcessor):
 
     def __init__(self, min_length: int, eos_token_id: Union[int, List[int]]):
         if not isinstance(min_length, int) or min_length < 0:
-            raise ValueError(f"`min_length` has to be a positive integer, but is {min_length}")
+            warnings.warn(f"`min_length` has to be a positive integer, but is {min_length}")
```
Comment (sgugger) [inline src/transformers/generation/logits_process.py:110]: Here what would be the goal of setting a `min_length` to a negative value? (nit: the error message should say non-negative by the way).

File: src/transformers/generation/logits_process.py [hunk 2]
```diff
@@ -106,12 +107,12 @@ class MinLengthLogitsProcessor(LogitsProcessor):
...
 
         if isinstance(eos_token_id, int):
             eos_token_id = [eos_token_id]
         if not all([isinstance(i, int) for i in eos_token_id]) or any([i < 0 for i in eos_token_id]):
-            raise ValueError(f"`eos_token_id` has to be a list of positive integers, but is {eos_token_id}")
+            warnings.warn(f"`eos_token_id` has to be a list of positive integers, but is {eos_token_id}")
```
Comment (sgugger) [inline src/transformers/generation/logits_process.py:115]: User `logger.warn` here and below. (`logger` is already defined in this file normally).

**Review by sgugger — COMMENTED** [review]
Thanks for working on this, just one comment and one question!

File: src/transformers/generation/logits_process.py (hunk 1 above)
Comment (lewtun) [reply src/transformers/generation/logits_process.py]: Ah yes, I thought this was needed to enable open-ended generation but I just checked and it's not the case. Now reverted back to raising an error.

**Review by sgugger — APPROVED** [review]

**Pull request merged**